### PR TITLE
Algorithm updates

### PR DESCRIPTION
**Pull request by HariniSundaram0**
# Description:
This pull request focuses on creating and saving data between sessions. The spotify genre seeds and user preferences are saved using UserDefaults, and the cache is saved to disk. Users are able to select genres to remove from the feed.

# Milestones:
P2: 
- Users can play songs via the ‘liked songs’ / ‘Library’ View
- Users are able to listen to songs without too long of a lag (persisting between user sessions).

# Test Plan:
Here is a test print statement showing results from the persisted genre preferences and cache. 
Screen Shot 2022-08-01 at 5 12 25 PM

As you can see, a dropdown exists: 
IMG_9189

### CODE REVIEW

File: Project-Hotpot/Project-Hotpot/Algorithm/Data Structures/Cache.swift [hunk 1]
```diff
@@ -10,12 +10,12 @@ import Foundation
...
+    let genreKeys: [String]
     
     init(genreKeys: [String]) {
-        for genreKey in genreKeys {
-            //manually initialize key to empty array
-            cache[genreKey] = []
-        }
+        self.genreKeys = genreKeys
+        super.init()
+        loadCache()
```
Comment (harshsinha) [inline Project-Hotpot/Project-Hotpot/Algorithm/Data Structures/Cache.swift:18]: Does this affect app startup or is there a way to init lazily? cc @toby1236

File: Project-Hotpot/Project-Hotpot/Algorithm/Data Structures/SongDetails.swift [hunk 2]
```diff
@@ -6,27 +6,12 @@
 //
 
 import Foundation
-
-class SongDetails: NSObject{
-    static public var numericalAttributes = ["danceability","energy","tempo","key"]
-    
+//chose not to wrap it within a class to prevent unneccessary decoding/encoding
+struct SongDetails: Codable, Hashable {
```
Comment (harshsinha) [inline Project-Hotpot/Project-Hotpot/Algorithm/Data Structures/SongDetails.swift:10]: Nice!

File: Project-Hotpot/Project-Hotpot/Managers/CacheManager.swift [hunk 3]
```diff
@@ -103,7 +103,7 @@ class CacheManager: NSObject {
                         //throws error later on
                         return nil
                     }
-                    return SongDetails(uri: uri, id: id, danceability: danceability, energy: energy, tempo: tempo, key: key)
+                    return SongDetails(uri: uri, id: id, danceability: Float(danceability), energy: Float(energy), tempo: Float(tempo), key: Float(key))
```
Comment (harshsinha) [inline Project-Hotpot/Project-Hotpot/Managers/CacheManager.swift:106]: Float causes some weird math and rounding issues, maybe use double (https://www.codingem.com/double-vs-float-swift/)

File: Project-Hotpot/Project-Hotpot/Managers/GenreQueueManager.swift [hunk 4]
```diff
@@ -29,9 +29,8 @@ class GenreQueueManager: NSObject {
     }
     
     func reinitializeQueue () {
-        if(self.genreQueue.isEmpty) {
-            let userGenres = UserSettingsManager.shared().userGenres
-            self.genreQueue.enqueueFromList(genres: Array(userGenres).shuffled())
+        if (self.genreQueue.isEmpty) {
+            self.genreQueue.enqueueFromList(genres: Array(UserSettingsManager.shared().userGenres).shuffled())
```
Comment (harshsinha) [inline Project-Hotpot/Project-Hotpot/Managers/GenreQueueManager.swift:33]: Does this get called every time user prefs change?

File: Project-Hotpot/Project-Hotpot/ViewControllers/HomeViewController.swift [hunk 5]
```diff
@@ -117,48 +104,37 @@ class HomeViewController: ViewController {
...
         }
     }
     
     //TODO: add completion block -> have to manually move card a little to re-reset card.
     func resetSong(completion: @escaping (_ result: Result<Void, Error>) -> Void) {
         let algInstance = SongAlgorithm()
-        algInstance.playNewSong { result in
-            completion(result)
-        }
+        algInstance.getAlgorithmSong { result in
```
Comment (harshsinha) [inline Project-Hotpot/Project-Hotpot/ViewControllers/HomeViewController.swift:127]: nit: Calling it "getNextSong" could make it more clear

File: Project-Hotpot/Project-Hotpot/ViewControllers/SettingsViewController.swift [hunk 6]
```diff
@@ -6,16 +6,55 @@
...
+            guard let removeGenres = self?.genresToRemove,
+                  let addGenres = self?.genresToAdd
+            else {
+                NSLog("returning, unable to access temporary arrays")
+                return
+            }
+            let removeDifference = removeGenres.subtracting(addGenres)
+            let addDifference = addGenres.subtracting(removeGenres)
+            UserSettingsManager.shared().removeMultipleGenresFromPreferences(genres: removeDifference)
+            UserSettingsManager.shared().addMultipleGenresFromPreferences(genres: addDifference)
```
Comment (harshsinha) [inline Project-Hotpot/Project-Hotpot/ViewControllers/SettingsViewController.swift:50]: addMultipleGenres**To**Preferences?

**Review by harshsinha — COMMENTED** [review]
Good job addressing all previous feedback. Just some nit comments :-)

File: Project-Hotpot/Project-Hotpot/Managers/UserSettingsManager.swift [hunk 7]
```diff
@@ -71,7 +71,14 @@ class UserSettingsManager: NSObject {
         }
         removedGenres.insert(genre)
         userGenres.remove(genre)
-        
+    }
+    
+    func removeMultipleGenresFromPreferences(genres: Set<String>) {
+        genres.map{ genre in removeGenreFromPrefences(genre: genre)}
```
Comment (toby1236) [inline Project-Hotpot/Project-Hotpot/Managers/UserSettingsManager.swift:78]: Nit: `genres.forEach` (https://developer.apple.com/documentation/swift/array/foreach(_:))
Also, another good opportunity to use `$0`

File: Project-Hotpot/Project-Hotpot/Algorithm/SongAlgorithm.swift [hunk 8]
```diff
@@ -41,9 +41,9 @@ class SongAlgorithm{
     }
     
     func getRandomGenre() -> String? {
-        //returns nil if the queue is empty
+        //returns nil if the queue somehow returns a nil object, which cannot happen
```
Comment (toby1236) [inline Project-Hotpot/Project-Hotpot/Algorithm/SongAlgorithm.swift:44]: Hmm... why does the signature of `GenreManager`'s `getGenre` return an optional if that value cannot be `nil`?

Non-optional variables are very powerful, they allow you to rely on having a value without requiring unnecessary nil checks to guard against NPEs.

File: Project-Hotpot/Project-Hotpot/Algorithm/Data Structures/Cache.swift (hunk 1 above)
Comment (toby1236) [reply Project-Hotpot/Project-Hotpot/Algorithm/Data Structures/Cache.swift:18]: Swift does have the concept of lazy: https://www.hackingwithswift.com/example-code/language/what-are-lazy-variables

It looks like this is synchronously loading from disk. The payloads are so small that the benefit of doing this lazily or in the background is likely unnoticeable to the user.

But it is worth pointing out, it's best practice to do all disk-based operations async. This ensures your UI doesn't stutter as cache and document size increases.

File: Project-Hotpot/Project-Hotpot/Algorithm/SongAlgorithm.swift (hunk 8 above)
Comment (toby1236) [reply Project-Hotpot/Project-Hotpot/Algorithm/SongAlgorithm.swift:44]: (note, this is also why I dislike `!` so much: it breaks this contract and robs non-optional variables of their power to abolish unnecessary NPE checks. Thank you for steering clear of `!`!)

**Pull request merged**